### ISSUE TITLE
azure-native allows resource-groups with duplicate names, then breaks

### ISSUE THREAD

**Issue by devnev** [pulumi/pulumi#16597]
### What happened?

I was trying to create two resource groups in Azure, in a loop over resource group names. I accidentally created resource groups with the same Name field - different pulumi IDs though - by using the wrong variable. `up` succeeded but in azure only one resource group was created. I fixed the variable in the code, so each group would be defined with the correct distinct name. In my case this had the effect that `up` renamed the existing group to the other name. After this, resource changes were failing with Resource Group not found, as they were referring to the resource group name which now no longer existed. I tried a refresh, but this actually made things worse, as after the refresh I was getting a snapshot integrity failure for all commands. The integrity failure was from references to the resource group that `refresh` had deleted from the state because it had actually been renamed to the other resource group. I ended up editing the state file to change the URN of the one resourcegroup that actually existed to the one being referenced.

### Example

Initial resources with mistake:

```go
resourceGroup1, err := resources.NewResourceGroup(ctx, `resource-group-1`, &resources.ResourceGroupArgs{
	ResourceGroupName: "Group1",
	Location: pulumi.StringPtr("westus2"),
})
resourceGroup2, err := resources.NewResourceGroup(ctx, `resource-group-2`, &resources.ResourceGroupArgs{
	ResourceGroupName: "Group1", //<--mistake
	Location: pulumi.StringPtr("westus2"),
})
storageAccount, err := storage.NewStorageAccount(ctx, "storage-account", &storage.StorageAccountArgs{
	AccountName:       pulumi.StringPtr("pulumi-bug-storage-account"),
	ResourceGroupName: resourceGroup1
	Kind:              storage.KindStorageV2,
	Sku: storage.SkuArgs{
		Name: storage.SkuName_Standard_ZRS,
	},
})
```

run pulumi up
then change the line with //<--mistake to "Group2"
run pulumi up - fails with ResourceGroup not found
run pulumi refresh - ok
run pulumi up - fails with integrity failure

as I'm not sure if the behaviour is deterministic, names and references might need swapping around to reproduce

### Output of `pulumi about`

```
CLI
Version      3.122.0
Go Version   go1.22.4
Go Compiler  gc

Host
OS       darwin
Version  14.5
Arch     x86_64

Backend
Name           [redacted]
URL            file://~
User           [redacted]
Organizations
Token type     personal

Pulumi locates its logs in /var/folders/sz/rcyd6rg51cbb4qzqsvzyn92m0000gn/T/ by default
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by Frassle** [comment]
This looks similar to https://github.com/pulumi/pulumi/issues/9925, but the integrity errors are probably their own unique bug that needs checking out.

**Comment by lunaris** [comment]
Hi @devnev, I know you mentioned you were dubious about how reproducible this is -- I don't suppose you have a program/configuration that consistently yields snapshot integrity violation errors, do you? For the record, here's what I can reproduce (on latest versions of both the Pulumi CLI and the various SDKs):

* `pulumi up` with duplicate names -- succeeds. As @Frassle suggests, behind the scenes, Azure has only created the one resource group, but this single group ends up being linked to both the Pulumi resources due to the second "create" call being returned the ID from the first.
* `pulumi up` after fixing the duplicate -- this too succeeds for me. Since `name` is a "replace triggering" property, Pulumi replaces the entire resource, which manifests as a deletion/recreation. The single group in Azure that links to both groups in Pulumi's state is deleted and a new one created in its place. The recreated resource group has a _different ID_ to its predecessor, which is reflected in the Pulumi state _only_ for the resource that changed. This means one resource is out of sync.
* `pulumi refresh` -- this too succeeds for me, though the effects are unfortunate as you say. Because the ID in Azure doesn't match that in the Pulumi state for the unchanged resource, Pulumi thinks there is no match for it and proceeds to delete its copy of that resource.
* `pulumi up` -- this then successfully recreates the resource we unfortunately deleted.

So yes, definitely some undesirable behaviour there due to Azure's implementation of idempotent create calls (that's what #9925 linked by @Frassle is about), but I can't reproduce the snapshot issues unfortunately. If you do have a program that does (or even literal logs from when it did through for you), I'd love to check those out and see if we can resolve that part.

**Comment by lunaris** [comment]
I don't _know_ that it's the same issue, but I'm fairly confident that this is the same as #16667, which hopefully will be addressed by #16699. As above though, if you have a solid reproduction I'll definitely look into that too!
